facade/src/test/java/org/jboss/pnc/facade/providers/UserProviderImplTest.java

```diff
@@ -0,0 +1,180 @@
+/**
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2014-2019 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.pnc.facade.providers;
+
+import org.jboss.pnc.dto.response.Page;
+import org.jboss.pnc.facade.validation.InvalidEntityException;
+import org.jboss.pnc.model.User;
+
+import org.jboss.pnc.spi.datastore.repositories.UserRepository;
+import org.jboss.pnc.spi.datastore.repositories.api.Predicate;
+import org.jboss.pnc.spi.datastore.repositories.api.Repository;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class UserProviderImplTest extends AbstractProviderTest<User> {
+
+    @Mock
+    private UserRepository repository;
+
+    @InjectMocks
+    private UserProviderImpl provider;
+
+    private User userMock = prepareNewUser("tada");
+
+    @Override
+    protected AbstractProvider provider() {
+        return provider;
+    }
+
+    @Override
+    protected Repository<User, Integer> repository() {
+        return repository;
+    }
+
+    @Before
+    public void setup() {
+        List<User> users = new ArrayList<>();
+        users.add(prepareNewUser("boris"));
+        users.add(prepareNewUser("theresa"));
+        users.add(prepareNewUser("david"));
+        users.add(userMock);
+
+        fillRepository(users);
+    }
+
+    @Test
+    public void testGetExistingUser() {
+        // Prepare
+        User user = repositoryList.get(0);
+
+        when(repository.queryByPredicates(any(Predicate.class))).thenAnswer(inv -> user);
+
+        // When
+        org.jboss.pnc.dto.User userDTO = provider.getOrCreateNewUser("test");
```

Comment:
I know when it's so mocked it actually doesn't have an effect, but I think you should still ask for a username that you actually receive.
```suggestion
        org.jboss.pnc.dto.User userDTO = provider.getOrCreateNewUser(user.getUsername());
```